traits/trait_observers.py

```diff
@@ -0,0 +1,153 @@
+from types import MethodType
+import weakref
+
+from .trait_base import Uninitialized
+
+
+class BaseObserverEvent(object):
+    pass
+
+
+class ObserverEvent(BaseObserverEvent):
+
+    def __init__(self, object, name, old, new):
+        self.object = object
+        self.name = name
+        self.old = old
+        self.new = new
+
+
+class ListObserverEvent(BaseObserverEvent):
+
+    def __init__(self, object, name, old, new):
+        self.object = object
+        self.name = name
+        self.old = old
+        self.new = new
+        self.added = new.added
+        self.removed = new.removed
+        self.index = new.index
+
+
+class TraitObserverNotifier(object):
+    """ Observer for a trait on a HasTraits instance.
+
+    Attributes
+    ----------
+    observer : callable or WeakMethod
+        The observer callback or a weak reference to it if it is a method.
+    notifier_observer : method
+        The method to use for dispatch, either
+    owner : list of callables
+        The list of notifiers that this is part of.
+    target : optional HasTraits instance
+        An optional object which is required for the observer to function.  If
+        provided, then only a weak reference is held to this object.
+    event_factory : callable
+        A factory function or class that creates appropriate event objects.
+
+    Parameters
+    ----------
+    observer : callable
+        The observer callback.  This should accept a single ObserverEvent
+        instance as an argument.  If the callable is a method, only a weak
+        reference to the method will be used.
+    owner : list of callables
+        The list of notifiers that this is part of.
+    target : optional HasTraits instance
+        An optional object which is required for the observer to function.  If
+        provided, then only a weak reference is held to this object.
+    event_factory : callable
+        A factory function or class that creates appropriate event objects.
+    """
+
+    def __init__(self, observer, owner, target=None, event_factory=ObserverEvent):
+        if isinstance(observer, MethodType):
+            # allow observing object methods to be garbage collected
+            self.observer = weakref.WeakMethod(observer, self.observer_deleted)
```

Comment:
It's inconsistent that we keep a weak reference for a method, but a strong reference for a function. That's baked in to the old system, and would be very difficult to change there, but we should consider whether we want to fix that inconsistency in the new system rather than propagating it.

Reply:
Also: we need to be very careful indeed when using weakref callbacks. If we need finalization, weakref.finalize may be a better option. Or we might consider how bad it would be not to use weakrefs at all.